.coveragerc

```diff
@@ -1,4 +1,9 @@
 [run]
 parallel = True
 data_file = /tmp/${USER}/cov/coverage.db
-omit = */site-packages/*
+source =
+    lib/pbench
+omit =
+    */site-packages/*
+    /tmp/${USER}/tox/*/bin/bottle.py
+    /tmp/${USER}/tox/*/lib/python3*
```

Comment:
I could use some help with the context, here.

Before, we were testing everything under the repo root, and now we're going to test only the stuff under `lib/pbench`?  (That seems like a narrowing of focus -- how does it now find **more** stuff which is untested?)

Starting at `lib/pbench`, how does it end up looking in places like `/tmp` such that we need to exclude them?  Does it have other (implicit) start locations?  Is there something in the code that we're having it test which reaches out to `/tmp`?  (E.g., should we be excluding `lib/pbench/test` -- we don't want to include the test coverage machinery in the code being coverage tested...right?)

Reply:
>  does not include command line interface files without `.py` extension.

I missed the directive for this...is it on the command line or in some other configuration file?  (Or, is this part of the `site-packages` exclusion?)

Reply:
> I could use some help with the context, here.

Me too - starting with how to run the thing... I'm almost certainly missing a bunch of things, but grepping through the tree looking for `coverage` finds precious little. Help!

Reply:
> Before, we were testing everything under the repo root, and now we're going to test only the stuff under `lib/pbench`? (That seems like a narrowing of focus -- how does it now find **more** stuff which is untested?)

The `source` statement does not mean a narrowing of what is considered for coverage, but the opposite: it directs the coverage.py utility to inspect all the modules found under that tree and record what was found, in addition to modules executed (and not in the list of omissions), so that one can see what was covered or not for the entire library, not just those modules that happen to have some part of the tested.

> Starting at `lib/pbench`, how does it end up looking in places like `/tmp` such that we need to exclude them? Does it have other (implicit) start locations? Is there something in the code that we're having it test which reaches out to `/tmp`? (E.g., should we be excluding `lib/pbench/test` -- we don't want to include the test coverage machinery in the code being coverage tested...right?)

This is mostly because of how `tox` is directed to use `/tmp/${USER}` a [line 6](https://github.com/distributed-system-analysis/pbench/blob/main/tox.ini#L6) of `tox.ini` (when `TOXWORKDIR` is not pre-defined).  So all pip installs land under that `/tmp/${USER}` hierarchy.

> > does not include command line interface files without `.py` extension.
> 
> I missed the directive for this...is it on the command line or in some other configuration file? (Or, is this part of the `site-packages` exclusion?)
> 
> Or is it because the tool has [always only measured files with the `.py` extension](https://coverage.readthedocs.io/en/coverage-5.0/source.html#execution)?

It is not clear this is about what coverage measures or not: in the absence of any `source`, `include`, or `omit` directives, coverage will record functions executed in any python code it is found.

This is about what additional data surrounding the executable code is included in addition to what is executed, and what should be ignored or not.

So when you specify `source` and a directory, it only looks for `*.py` files in that hierarchy.  If you want to explicitly include files without a `.py` extension, then you have to specify them individually IIF they are not executed.

Reply:
> > I could use some help with the context, here.
> 
> Me too - starting with how to run the thing... I'm almost certainly missing a bunch of things, but grepping through the tree looking for `coverage` finds precious little. Help!

Coverage data is always collected in `/tmp/${USER}/cov` for a given `tox` environment (if that environment executes python3 code).

To see that coverage, you have to either review the generated HTML files (see below), or run the `coverage` command in the top-level of the pbench source tree (where the `.coverage` file is located).

You can find the output of the `tox` tests what we are doing today, e.g. here is the output from a `tox -e py3-agent`:

```
py3-agent run-test: commands[0] | pytest --cov --cov-append --cov-report html:/tmp/${USER}/tox/py3-agent/cov/html ./lib/pbench/test/unit/common ./lib/pbench/test/unit/agent ./lib/pbench/test/functional
```

And near the end of the `tox` output you'll see:

```
----------- coverage: platform linux, python 3.8.6-final-0 -----------
Coverage HTML written to dir /tmp/${USER}/tox/py3-agent/cov/html
```

Reply:
Regarding @webbnh's comments, yes, you are correct that `source` narrows the modules measured so that the `omit` statements are not needed.